### PR TITLE
Change cartesian product notification input position

### PR DESCRIPTION
**Pull request by Mats-SX**
A cartesian product notification will now use the input position of its
wrapping clause (eg, a MATCH) as well as point to the identifiers that are involved
in causing the cartesian product to happen in its description.

### CODE REVIEW

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/ast/Clause.scala [hunk 1]
```diff
@@ -96,20 +96,19 @@ case class Match(optional: Boolean, pattern: Pattern, hints: Seq[UsingHint], whe
...
+      }.foldLeft(intermediateState) { (innerState, identifiers) =>
+        innerState.addNotification(CartesianProductNotification(position, identifiers))
       }
 
-      if (rest.nonEmpty) loop(rest.head, rest.tail)
+      if (rest.nonEmpty) loop(rest.head, rest.tail, updatedState) else updatedState
     }
 
-    if (nodes.nonEmpty) loop(nodes.head, nodes.tail)
+    val finalState = if (nodes.nonEmpty) loop(nodes.head, nodes.tail, state) else state
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/ast/Clause.scala:109]: Why this change? When does this make a difference?

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/SemanticState.scala [hunk 2]
```diff
@@ -186,7 +187,7 @@ import org.neo4j.cypher.internal.compiler.v2_3.SemanticState.ScopeLocation
 case class SemanticState(currentScope: ScopeLocation,
                          typeTable: ASTAnnotationMap[ast.Expression, ExpressionTypeInfo],
                          recordedScopes: ASTAnnotationMap[ast.ASTNode, Scope],
-                         notificationLogger: InternalNotificationLogger = devNullLogger) {
+                         notifications: Set[InternalNotification] = Set.empty) {
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/SemanticState.scala:190]: This changes so that we always record notifications even in the case when we use `devNullLogger`. I am not seeing what this change has to do with positioning?

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/ast/Clause.scala (hunk 1 above)
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/ast/Clause.scala:109]: The difference is that the `SemanticState` previously contained a mutable logger field, but now contains an immutable `Set[InternalNotification]`. Due to that, we can't just add a notification to the state, but we need to keep the return value from the `innerState.addNotification` call and send the state with all notifications back on the last line.

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/SemanticState.scala (hunk 2 above)
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/SemanticState.scala:190]: `SemanticState` is a monad and should not contain mutable state (the InternalNotificationLogger is mutable). @boggle and I made this discovery and changed it while we were passing through.

The `devNullLogger` still has its use on line 31 in `SemanticChecker` above, where notifications are passed to the logger. If that's the `devNullLogger`, notifications will be discarded.

File: community/kernel/src/test/java/org/neo4j/graphdb/impl/notification/NotificationCodeTest.java [hunk 3]
```diff
@@ -45,4 +49,24 @@ public void shouldConstructNotificationFor_INDEX_HINT_UNFULFILLABLE()
...
+        assertThat( notification.getSeverity(), equalTo( SeverityLevel.WARNING ) );
+        assertThat( notification.getCode(), equalTo( "Neo.ClientNotification.Statement.CartesianProduct" ) );
+        assertThat( notification.getPosition(), equalTo( InputPosition.empty ) );
+        assertThat( notification.getDescription(), equalTo( "If a part of a query contains multiple disconnected patterns, this will build a cartesian product " +
+                                                            "between all those parts. This may produce a large amount of data and slow down query processing. While " +
+                                                            "occasionally intended, it may often be possible to reformulate the query that avoids the use of this cross " +
+                                                            "product, perhaps by adding a relationship between the different parts or by using OPTIONAL MATCH " +
+                                                            "(identifiers are: n at (0, 1, 3), node2 at (2, 3, 4))" ) );
+    }
```
Comment (boggle) [inline community/kernel/src/test/java/org/neo4j/graphdb/impl/notification/NotificationCodeTest.java]: I find that printing the input positions like that is not helpful, actually I'm not sure that including them in the details string is helpful at all. Why not rather include the sets of connected identifiers? Something like: "(first connected component is n, a, second connected component is m, b)"

Comment (Mats-SX) [inline community/kernel/src/test/java/org/neo4j/graphdb/impl/notification/NotificationCodeTest.java]: That's not quite how it works. We have one warning per disconnected component (counting in written order, meaning that `MATCH a, b-->c, c-->d, d-->e` will report 3 notifications with identifier pairs (b,c), (c,d), (d,e), and not just (a)), which means each warning only knows about its identifiers.

But I could make it not print input positions of course.

**Comment by pontusmelke** [comment]
Note this does not solve the issue of reported position using when using `EXPLAIN` or `PROFILE`. This work will be addressed in a separate PR
